compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1653,8 +1654,157 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ),
                 Mismatch::Fixed(s) => (s.into(), s.into(), None),
             };
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+
+            enum Similar<'tcx> {
+                Adts(ty::AdtDef<'tcx>, ty::AdtDef<'tcx>),
+                PrimitiveFound(Ty<'tcx>, ty::AdtDef<'tcx>),
+                PrimitiveExpected(ty::AdtDef<'tcx>, Ty<'tcx>),
+            }
+
+            let primitive_sym = |kind: &_| match kind {
+                ty::Bool => Some(sym::bool),
+                ty::Char => Some(sym::char),
+                ty::Float(f) => match f {
+                    ty::FloatTy::F32 => Some(sym::f32),
+                    ty::FloatTy::F64 => Some(sym::f64),
+                },
+                ty::Int(f) => match f {
+                    ty::IntTy::Isize => Some(sym::isize),
+                    ty::IntTy::I8 => Some(sym::i8),
+                    ty::IntTy::I16 => Some(sym::i16),
+                    ty::IntTy::I32 => Some(sym::i32),
+                    ty::IntTy::I64 => Some(sym::i64),
+                    ty::IntTy::I128 => Some(sym::i128),
+                },
+                ty::Uint(f) => match f {
+                    ty::UintTy::Usize => Some(sym::usize),
+                    ty::UintTy::U8 => Some(sym::u8),
+                    ty::UintTy::U16 => Some(sym::u16),
+                    ty::UintTy::U32 => Some(sym::u32),
+                    ty::UintTy::U64 => Some(sym::u64),
+                    ty::UintTy::U128 => Some(sym::u128),
+                },
+                _ => None,
+            };
```

Comment:
This should be an associated function for `TyKind`. Should be okay for now, because it can be in a followup.

Reply:
I don't think it can be an associated function on TyKind (or Symbol for that matter) because that leads to cyclic dependencies. But I'm happy to make it a free function/trait somewhere, or an associated function on `rustc_middle::ty::Ty`